### PR TITLE
Feature/mail receipt return

### PR DESCRIPTION
**Pull request by cableman**

### CODE REVIEW

File: plugins/notification/notification.js [hunk 1]
```diff
@@ -492,49 +491,64 @@ Notification.prototype.checkInReceipt = function checkInReceipt(mail, items, lan
     patrons: []
   };
 
-  // Build promise with patron information.
-  var patrons = [];
-  for (var username in items) {
-    var promise = this.getPatronInformation(username);
-    patrons.push(promise);
+  console.log('HEHE1');
```
Comment (tuj) [inline plugins/notification/notification.js]: Øhh

File: plugins/notification/notification.js [hunk 2]
```diff
@@ -492,49 +491,64 @@ Notification.prototype.checkInReceipt = function checkInReceipt(mail, items, lan
...
+
+  var patronsInformation = JSON.parse(JSON.stringify(items.patronsInformation));
+  delete items.patronsInformation;
```
Comment (tuj) [inline plugins/notification/notification.js]: Why remove?

File: plugins/notification/notification.js [hunk 3]
```diff
@@ -492,49 +491,64 @@ Notification.prototype.checkInReceipt = function checkInReceipt(mail, items, lan
...
+        patronIdentifier: patronIdentifier,
+        name: patronInformation.hasOwnProperty('personalName') ? patronInformation.personalName : 'Unknown',
+        fines: layout.fines ? self.renderFines(mail, patronInformation.fineItems) : '',
+        loans: layout.loans ? self.renderLoans(mail, 'receipt.loans.headline', patronInformation.chargedItems, patronInformation.overdueItems) : '',
+        reservations: layout.reservations ? self.renderReservations(mail, patronInformation.unavailableHoldItems) : '',
+        reservations_ready: layout.reservations_ready ? self.renderReadyReservations(mail, patronInformation.holdItems) : '',
+        check_ins: layout.check_ins ? self.renderCheckIn(mail, items[patronInformation.patronIdentifier]) : ''
+      };
+
+      context.patrons.push(JSON.parse(JSON.stringify(content)));
```
Comment (tuj) [inline plugins/notification/notification.js]: Is this necessary? If it is comment.

File: plugins/notification/notification.js [hunk 4]
```diff
@@ -492,49 +491,64 @@ Notification.prototype.checkInReceipt = function checkInReceipt(mail, items, lan
...
   };
 
-  // Build promise with patron information.
-  var patrons = [];
-  for (var username in items) {
-    var promise = this.getPatronInformation(username);
-    patrons.push(promise);
+  console.log('HEHE1');
+
+  var patronsInformation = JSON.parse(JSON.stringify(items.patronsInformation));
```
Comment (tuj) [inline plugins/notification/notification.js]: Is this necessary? If it is comment.

File: plugins/notification/notification.js [hunk 5]
```diff
@@ -492,49 +491,64 @@ Notification.prototype.checkInReceipt = function checkInReceipt(mail, items, lan
...
+      context.patrons.push(JSON.parse(JSON.stringify(content)));
+    }
+    else {
+      // Unknown patrons (item was not checkout before check-in etc.).
+      content = {
+        name: 'Unknown user',
+        check_ins: layout.check_ins ? self.renderCheckIn(mail, items['unknown']) : ''
+      };
+
+      context.patrons.push(JSON.parse(JSON.stringify(content)));
```
Comment (tuj) [inline plugins/notification/notification.js]: Is this necessary? If it is comment.

File: plugins/notification/notification.js [hunk 6]
```diff
@@ -492,49 +491,64 @@ Notification.prototype.checkInReceipt = function checkInReceipt(mail, items, lan
...
-          deferred.reject(data);
-          // End processing.
-          return;
-        }
+  // Render receipt.
+  var result = '';
+  if (mail) {
+    // The context is optimized for print receipt, so we need to change it to
+    // send a mail to each patron and not one big to one mail patron.
+    var patrons = JSON.parse(JSON.stringify(context.patrons));
```
Comment (tuj) [inline plugins/notification/notification.js:535]: Please look through all these, and make sure they are necessary

File: plugins/notification/notification.js [hunk 7]
```diff
@@ -492,49 +491,64 @@ Notification.prototype.checkInReceipt = function checkInReceipt(mail, items, lan
...
-      }
-    });
+      result = self.mailTemplate.render(data);
 
-    // Render receipt.
-    var result = self.printTemplate.render(context);
+      // Remove empty lines (from template engine if statements).
+      result = result.replace(/(\r\n|\r|\n){2,}/g, '$1\n');
+
+      self.sendMail(patronsInformation[patron.patronIdentifier].emailAddress, result).then(function () {
```
Comment (tuj) [inline plugins/notification/notification.js]: anonymous function??     maybe function success() {

File: public/js/services/receiptService.js [hunk 8]
```diff
@@ -40,7 +40,36 @@ angular.module('BibBox').service('receiptService', ['$q', 'tmhDynamicLocale', 'p
...
+     *
+     * @param {array} patronIdentifiers
+     *   The identifiers for the patrons.
+     *
+     * @returns {Function}
+     */
+    this.getPatronsInformation = function getPatronsInformation(patronIdentifiers) {
+      var deferred = $q.defer();
+
+      proxyService.once('notification.response', function (mailAddresses) {
```
Comment (tuj) [inline public/js/services/receiptService.js:53]: No unique id?

**Review by tuj — CHANGES_REQUESTED** [review]

File: plugins/notification/notification.js (hunk 2 above)
Comment (cableman) [reply plugins/notification/notification.js]: Yes, as the receipt builder loops over all elements in items.

File: public/js/services/receiptService.js (hunk 8 above)
Comment (cableman) [reply public/js/services/receiptService.js:53]: No this is how all the notifications are made. It's outside this PR's scope to re-do that

File: plugins/notification/notification.js (hunk 1 above)
Comment (cableman) [reply plugins/notification/notification.js]: Fixed

File: plugins/notification/notification.js (hunk 4 above)
Comment (cableman) [reply plugins/notification/notification.js]: Fixed

File: plugins/notification/notification.js (hunk 2 above)
Comment (cableman) [reply plugins/notification/notification.js]: Comment added to code as well

File: plugins/notification/notification.js (hunk 3 above)
Comment (cableman) [reply plugins/notification/notification.js]: Comment added

File: plugins/notification/notification.js (hunk 5 above)
Comment (cableman) [reply plugins/notification/notification.js]: Comment added

File: plugins/notification/notification.js (hunk 6 above)
Comment (cableman) [reply plugins/notification/notification.js:535]: Comment added.

File: plugins/notification/notification.js (hunk 7 above)
Comment (cableman) [reply plugins/notification/notification.js]: done

**Pull request merged**